utils.py

```diff
@@ -0,0 +1,169 @@
+from functools import partial
+from multiprocessing import Pool, cpu_count
+
+import numpy as np
+from pyproj import Transformer
+from shapely import ops, wkb
+
+TRANSFORMERS = {}
+
+
+def get_epsg(geom):
+    if len(geom.geoms) == 1:
+        point = geom.centroid
+    else:
+        # antimeridian case
+        # Transform to the first UTM zone, get the centroid, then transform back
+        # Doesnt matter if it's the right EPSG; just need it to be in
+        # projected coordinates to make the centroid easy
+        utm_epsg = 32601 if geom.centroid.y > 0 else 32701
+        src, dst = 4326, utm_epsg
+        point = transform(transform(geom, src, dst).centroid, dst, src)
```

Comment:
that is a very pleasing figure haha. how'd you make it?


I think that even though it would be terrible to warp normal points from all over the globe into UTM 1N, doing the `warp -> centroid -> inverse warp` shouldn't distort it too much? I'm gonna draw myself a figure since i'm unsure, but i'll double check and compare to the EPSG codes from the current version of your DB

Reply:
well i've confused myself, since it seems like the warping might not necessarily get the right centroid if it's a really nonlinear bad warping... but I'm not sure which (if any) bursts it messed it up for?

burst_id_jpl,EPSG,epsg
t001_000110_iw1,32631,32630
t001_000111_iw1,32631,32630
t001_000112_iw1,32631,32630
t001_000113_iw1,32631,32630

```python
import pandas as pd
import sqlite3
import os

os.chdir("/home/staniewi/dev")
db1 = "burst_map_IW_000001_375887.OPERA-JPL.20220818_113416.sqlite3" # made from original version
db2 = "burst_map_margin3000.sqlite3"        # from new version

with sqlite3.connect(db1) as con:
    df1 = pd.read_sql("select * from burst_id_map", con)
with sqlite3.connect(db2) as con:
    df2 = pd.read_sql("select * from burst_id_map", con)

mismatches = df1.EPSG != df2.epsg

print(df1.shape, df2.shape)  # (1127661, 13) (1127661, 13)
print(mismatches.sum())      # 104617


df_merged = pd.merge(df1.loc[mismatches, ['burst_id_jpl','EPSG', ]], df2.loc[mismatches, ['epsg', 'burst_id_jpl']], on="burst_id_jpl")
df_merged.to_csv("~/dev/burst_diffs.csv", index=False)
```

Looking at the mismatches, a lot are just off by one, which might be either bursts on the borderline of two UTM zones getting a slightly different centroid, or might be your keeping adjacent subswaths in the same UTM zone:

(burst id, then your version of the EPSG, then my script's output)
```
burst_id_jpl,EPSG,epsg
t001_000110_iw1,32631,32630
t001_000111_iw1,32631,32630
t001_000112_iw1,32631,32630
... (lots more like this)
t001_000448_iw1,3413,32627
t001_000449_iw1,3413,32627
(more on the borderline between polar and non-polar regions)
```

The biggest differences i've seen look like this:
```
t001_000669_iw2,32612,32601
t001_000669_iw3,32612,32660
t001_000670_iw1,32614,32601
t001_000670_iw2,32614,32601
```
where these bursts on the anti-meridian are off due to what you pointed out in #1 .

Reply:
Also regarding the distortion of polygon bursts on UTM1N (in the figure above): I was bit confused when thinking about it. I've checked that you apply "warp-centroid-warp_back" only for Antimeridian case i.e. multiple polygon in feature's geometry. Using UTM 1N would be okay in that case. Sorry for the false alarm.